Bigquery integration

gathineou opened the pull request.

hz-lschick commented on mara_db/dbs.py:
```diff
@@ -61,6 +61,21 @@ def __init__(self, host: str = None, port: int = None, database: str = None,
+        https://cloud.google.com/bigquery/docs/quickstarts/quickstart-client-libraries#before-you-begin
+
+        """
+        self.service_account_json = service_account_json
+        self.project_id = project_id
+        self.use_legacy_sql = use_legacy_sql
```
I miss here the option to set the default location and dataset; see as well: https://github.com/hz-lschick/mara-db-bigquery/blob/master/mara_db_bigquery/dbs.py

In addition, I suggest to name 'project_id' to 'project'. This just makes the name straight-forward ...

gathineou replied:
Yes, sounds good

hz-lschick commented on mara_db/shell.py:
```diff
@@ -79,6 +78,16 @@ def __(db: dbs.RedshiftDB, timezone: str = None, echo_queries: bool = None):
+    assert all(v is None for v in [timezone, echo_queries]), f"unimplemented parameter for BigQueryDB"
+
+    return ('bq query'
+            + (f' --use_legacy_sql=' + ('true' if db.use_legacy_sql else 'false'))
+            + ' --quiet --headless -n=1000000000 ')
```
Is there a need for quiet, headless or the 'n' argument?

in addition, I think we should pass here over the project id, default location and dataset id, see also https://github.com/hz-lschick/mara-db-bigquery/blob/master/mara_db_bigquery/shell.py

gathineou replied:
Yes, this can be improved. 
**--headless** Relatively minor because it doesn't seem to affect but I would keep for disabling any kind of user-interaction mode based on the description in the official docs (https://cloud.google.com/bigquery/docs/reference/bq-cli-reference)
**--quiet** similar concept. It doesn't seem to affect on some simple testing that I tried but I would keep since we are manipulating data through stdin/out and we don't want any metadata there.
About the **n** argument, I couldn't find another way to disable the default limits and I have this ugly thing. We need to explore more there and do it in a proper way. 
How do you actually do that? Please share if you more insights on this!

hz-lschick replied:
Headless and Quite sounds good to me. About **n**: The question is what do you want to do?; e.g. what command is running into the limit? Did you look at `bq load` or `bq extract`?

When this is really necessary, I would consider rather building an extension class to mara_pipelines.commands.sql.ExecuteSQL, e.g. 'ExecuteBigQuerySql' which has an additional parameter which is then passed over to 'n' . But this would make sence rather in a separate repository.

gathineou replied:
From the cli docs on bq-query command:
```
--max_rows or -n
    An integer specifying the number of rows to return in the query results. The default value is 100
```
The problem is that the basic **bq query** command is under the default 100 limit and I can not find a way to have unlimited results there...

hz-lschick replied:
I currently just use query_command for executing DDL executions, so, I don't have this problem. When you want to execute huge chunks of data, I recommend using [`bq extract`](https://cloud.google.com/bigquery/docs/exporting-data) which has other limits.

I am fine with this parameter in general but think it should be somehow optional. Each query has its cost in BigQuery, so, I think when you go over a limit, this should be actively 'activated' and not just be in the mara BigQuery module implementation to avoid extensive costs by accident.

You can configure local default parameters for the `bq` cli with an config file which you place locally on your server. This could be the easy solution for you. See here: https://cloud.google.com/bigquery/docs/bq-command-line-tool#setting_default_values_for_command-line_flags

gathineou replied:
**bq extract** is a good idea but its limited only to work with the cloud storage and there might be use-cases that need the data locally. We should consider this option though.

There is no need to have such a user option/argument in the command itself. The limit can be specified in the query itself so the user has total control on that.
Although, if we want to have global settings for some BQ parameters (which I agree makes sense due to costs by accident etc), I like the solution with the default config file option where we can define/hide some of our defaults (starting from this ugly n value) :)

hz-lschick commented on mara_db/shell.py:
```diff
@@ -326,12 +335,35 @@ def __(db: dbs.RedshiftDB, target_table: str, csv_format: bool = None, skip_head
            + f'{query_command(db, timezone)} \\\n      --command="{sql}"\n\n' \
            + s3_delete_tmp_file_command
 
+
+@copy_from_stdin_command.register(dbs.BigQueryDB)
+def __(db: dbs.BigQueryDB, target_table: str, csv_format: bool = None, skip_header: bool = None,
+       delimiter_char: str = None, quote_char: str = None, null_value_string: str = None, timezone: str = None):
+    sql = 'bq load'
+    if csv_format:
+        sql += ' --source_format=CSV'
+    else:
+        sql += ' --source_format=NEWLINE_DELIMITED_JSON'
+    if skip_header:
+        sql += ' --skip_leading_rows=1'  # requires already created table
+    else:
+        sql += '  --autodetect'  # schema auto-detection for CSV and JSON data
+    if delimiter_char is not None:
+        sql += f" --field_delimiter='{delimiter_char}'"
+    if null_value_string is not None:
+        sql += f" --null_marker='{null_value_string}'"
+    if quote_char is not None:
+        sql += f" --quote='{quote_char}'"
+
+    return sql + ' ' + target_table
+
+
 # -------------------------------
 
 
 @multidispatch
 def copy_command(source_db: object, target_db: object, target_table: str,
-                 timezone = None, csv_format = None, delimiter_char = None) -> str:
+                 timezone=None, csv_format=None, delimiter_char=None) -> str:
```
don't know why this is done here, but, ok...; personally prefer to make such things in a separate pull-request.

hz-lschick commented on mara_db/shell.py:
```diff
@@ -83,6 +83,9 @@ def __(db: dbs.BigQueryDB, timezone: str = None, echo_queries: bool = None):
     echo_queries = None
     assert all(v is None for v in [timezone, echo_queries]), f"unimplemented parameter for BigQueryDB"
 
+    # One time activation of the service-account used is required
+    # todo: maybe as part of db initialization
+    # gcloud auth activate-service-account --key-file='path-to/service-account.json'
```
Can we move this into the README or into the documentation of the init-method of class BigQueryDB?

gathineou replied:
Yes, I was also thinking to move it in a more transparent place.

Thanks for the review @hz-lschick 
In addition, please note that this PR is still missing the **copy_to_stdout_command** generic implementation in a way that can be used without issues from the **copy_command** (i.e. from BQ to Postrgres)

gathineou commented:
Hi @hz-lschick,
Just review your latest changes in the branch. In general, they look good apart than the copy_from_stdin_command where I modified the default behaviour in case a user wants to run it without the google-cloud bucket option.
+ I adjusted the documentation.
For the last changes you can check this commit: https://github.com/mara/mara-db/commit/80c37b6d3e0497264dad929855f9bc965230272f

hz-lschick commented:
@gathineou Just questioning why we implement the local file import for BigQuery, but this is not done for Redshift [here](https://github.com/mara/mara-db/blob/master/mara_db/shell.py#L303-L327). But I don't know if that is even possible - and - I think this would be out-of-scope for this branch, so: LGTM
